### PR TITLE
Game layout

### PR DESCRIPTION
**Pull request by kyokidG**
Major game functionalities implemented. Hardly commented for easier code review end edit.

### CODE REVIEW

File: app/static/css/game.css [hunk 1]
```diff
@@ -20,3 +12,281 @@
...
+}
+
+#gravity-bars .bar {
+    width: 4%;
+    margin-left: 0.2%;
+    display: inline-block;
+    border: 0.5px solid grey;
+}
+
+.unu  { height: 9%; }
```
Comment (cristimihok) [inline app/static/css/game.css]: I still don't agree with this solution but if you really want to keep it like this then at least use english numbering

File: app/static/js/game.js [hunk 2]
```diff
@@ -0,0 +1,247 @@
+$(document).ready(function () {
+    startTimer();
```
Comment (cristimihok) [inline app/static/js/game.js]: This is a way to big chunk of code to be added on the `ready()` callback. Try extracting this into a function or multiple functions and files

File: app/static/js/game.js [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+    //initialize g - this will be done on the server side, but for now, we do it here
+    var g = ((Math.random()*10)+1).toFixed(2);
+
+    $(document).mousemove(function(e){
+        var distX = e.pageX - gameBox.offset().left - myBody.width();
+        var distY = Math.abs(e.pageY - gameBox.offset().top - gameBox.height())-myBody.height();
+
+        var angle = Math.atan2(distX,distY)*(180/Math.PI);
+        if(angle<0) {
+          angle=0;
```
Comment (cristimihok) [inline app/static/js/game.js]: `angle = 0`

File: app/static/js/game.js [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        }
+        if(angle>90) {
+          angle=90;
```
Comment (cristimihok) [inline app/static/js/game.js]: idem

File: app/static/js/game.js [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        }
+        myCannon.css({ "-webkit-transform": 'rotate(' + angle + 'deg)'});
```
Comment (cristimihok) [inline app/static/js/game.js]: As far as I know you only support Chrome type browsers.

File: app/static/js/game.js [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+        myCannon.css({ '-moz-transform': 'rotate(' + angle + 'deg)'});
+        myCannon.css({ 'transform': 'rotate(' + angle + 'deg)'});
+    });
+    gameBox.mousedown(function(e){
+        if(e.which==1) {
```
Comment (cristimihok) [inline app/static/js/game.js]: if (...)

File: app/static/js/game.js [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+            $('#power').show();
+            $('#power').css({
+                'left': e.pageX - gameBox.offset().left,
+                'right': 'auto',
+                'top': e.pageY-gameBox.offset().top-5,
+                'bottom': 'auto'});
```
Comment (cristimihok) [inline app/static/js/game.js]: Closing parentheses on a new line

File: app/static/js/game.js [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+        }
+    });
+    $(document).mouseup(function(e){
```
Comment (cristimihok) [inline app/static/js/game.js]: Document level event handlers don't have to be registered after the dome finished loading``

File: app/static/js/game.js [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+
+                //starting Position of the shot
+                //we need the angle of the cannon to compute the starting position
+                var distX = e.pageX - gameBox.offset().left;
+                var distY = Math.abs(e.pageY - gameBox.offset().top - gameBox.height());
+                var angle = Math.atan2(distX-myBody.width(),distY-myBody.height())*(180/Math.PI);
+                if(angle<0) {
+                  angle=0;
+                }
+                if(angle>90) {
```
Comment (cristimihok) [inline app/static/js/game.js]: Idem ^

File: app/static/js/game.js [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+                //power of the shot
+                var power = $('#current-power').width()/3;
+                $('#power').hide();
+
+                //starting Position of the shot
+                //we need the angle of the cannon to compute the starting position
+                var distX = e.pageX - gameBox.offset().left;
+                var distY = Math.abs(e.pageY - gameBox.offset().top - gameBox.height());
+                var angle = Math.atan2(distX-myBody.width(),distY-myBody.height())*(180/Math.PI);
+                if(angle<0) {
```
Comment (cristimihok) [inline app/static/js/game.js]: Idem ^

File: app/static/js/game.js [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+                //left mouse button pressed
+                //power of the shot
+                var power = $('#current-power').width()/3;
+                $('#power').hide();
+
+                //starting Position of the shot
+                //we need the angle of the cannon to compute the starting position
+                var distX = e.pageX - gameBox.offset().left;
+                var distY = Math.abs(e.pageY - gameBox.offset().top - gameBox.height());
+                var angle = Math.atan2(distX-myBody.width(),distY-myBody.height())*(180/Math.PI);
```
Comment (cristimihok) [inline app/static/js/game.js]: Separate function parameters with a `, ` combination

File: app/static/js/game.js [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+                if(angle<0) {
+                  angle=0;
+                }
+                if(angle>90) {
+                  angle=90;
+                }
+                //compute the actual starting Position of the missile
+                var xPos = myCannon.offset().left-1-gameBox.offset().left+1.3*angle;
```
Comment (cristimihok) [inline app/static/js/game.js]: Readability

File: app/static/js/game.js [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+                var yPos = Math.floor(myCannon.offset().top-1-gameBox.offset().top);
+                var startPosition = [xPos,yPos];
+
+                //angle of the shot
+                var dx = e.pageX - gameBox.offset().left;
+                var dy = Math.abs(e.pageY - gameBox.offset().top - gameBox.height());
+                var angle = Math.atan2(dx-myBody.width(),dy-myBody.height())*(180/Math.PI);
```
Comment (cristimihok) [inline app/static/js/game.js]: You seem to have done this calculation previously. Repetitive

File: app/static/js/game.js [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+                if(Math.atan2(dx-myBody.width(),dy-myBody.height())*(180/Math.PI)-90<0) {
+                    var angle = Math.abs(Math.atan2(dx-myBody.width(),dy-myBody.height())*(180/Math.PI)-90);
+                }
+                else {
+                    var angle = -(Math.atan2(dx-myBody.width(),dy-myBody.height())*(180/Math.PI)-90);
+                }
+                if(angle<0) {
```
Comment (cristimihok) [inline app/static/js/game.js]: Readability

File: app/static/js/game.js [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+
+                //angle of the shot
+                var dx = e.pageX - gameBox.offset().left;
+                var dy = Math.abs(e.pageY - gameBox.offset().top - gameBox.height());
+                var angle = Math.atan2(dx-myBody.width(),dy-myBody.height())*(180/Math.PI);
+                if(Math.atan2(dx-myBody.width(),dy-myBody.height())*(180/Math.PI)-90<0) {
+                    var angle = Math.abs(Math.atan2(dx-myBody.width(),dy-myBody.height())*(180/Math.PI)-90);
+                }
+                else {
+                    var angle = -(Math.atan2(dx-myBody.width(),dy-myBody.height())*(180/Math.PI)-90);
```
Comment (cristimihok) [inline app/static/js/game.js]: Readability

File: app/static/js/game.js [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+                }
+                if(angle<0) {
+                    angle=0;
+                }
+                if(angle>90) {
```
Comment (cristimihok) [inline app/static/js/game.js]: Readability

File: app/static/js/game.js [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+                    angle=90;
+                }
+
+                //call myShot - to shot projectile from my cannon
+                if(myCurrentCooldown.width()>=Math.floor($("#my-cooldown").width())) {
```
Comment (cristimihok) [inline app/static/js/game.js]: Readability

File: app/static/js/game.js [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+                    myMissile.css( { display: "inherit" });
```
Comment (cristimihok) [inline app/static/js/game.js]: Unnecessary space

File: app/static/js/game.js [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+                    myCurrentCooldown.css( { width: 0+"px" } );
```
Comment (cristimihok) [inline app/static/js/game.js]: Unnecessary spaces

File: app/static/js/game.js [hunk 20]
```diff
@@ -0,0 +1,247 @@
...
+        else {
+            if(secs%60<10) { timer.text(Math.round(secs/60) + ":0" + secs%60); }
+            else { timer.text(Math.round(secs/60) + ":" + secs%60); }
+        }
+    }, 1000);}
+
+    $(document).bind("contextmenu", function(event) {
+        event.preventDefault();
+    });
+});
```
Comment (cristimihok) [inline app/static/js/game.js]: https://google.github.io/styleguide/javascriptguide.xml

File: app/static/js/game.js [hunk 21]
```diff
@@ -0,0 +1,247 @@
...
+                //call myShot - to shot projectile from my cannon
+                if(myCurrentCooldown.width()>=Math.floor($("#my-cooldown").width())) {
+                    myMissile.css( { display: "inherit" });
+                    myCurrentCooldown.css( { width: 0+"px" } );
+                    $.when(myShot(startPosition, angle, power) ).then(
+                        function(status) {
+                            $("#enemy-current-health").css( { width: $("#enemy-current-health").width()-101 + "px" } )
+                            if($("#enemy-current-health").width()<=0) {
+                                warningMessage.show();
+                                warningMessage.text("Ai castigat!");
```
Comment (cristimihok) [inline app/static/js/game.js]: Swap these two. First you set the text after you show to avoid glitches

File: app/static/js/game.js [hunk 22]
```diff
@@ -0,0 +1,247 @@
...
+                            }
+                        },
+                        function( status ) {
+                            //console.log("miss");
```
Comment (cristimihok) [inline app/static/js/game.js]: Remove if not used

File: app/static/js/game.js [hunk 23]
```diff
@@ -0,0 +1,247 @@
...
+                        },
+                        function( status ) {
+                            //$( "body" ).append( status );
```
Comment (cristimihok) [inline app/static/js/game.js]: Remove if not used

File: app/static/js/game.js [hunk 24]
```diff
@@ -0,0 +1,247 @@
...
+                }
+                if(angle>90) {
+                    angle=90;
+                }
+
+                //call myShot - to shot projectile from my cannon
+                if(myCurrentCooldown.width()>=Math.floor($("#my-cooldown").width())) {
+                    myMissile.css( { display: "inherit" });
+                    myCurrentCooldown.css( { width: 0+"px" } );
+                    $.when(myShot(startPosition, angle, power) ).then(
```
Comment (cristimihok) [inline app/static/js/game.js]: Extract the anonymous callbacks in their own named functions. It is hard to follow this way.

File: app/static/js/game.js [hunk 25]
```diff
@@ -0,0 +1,247 @@
...
+                }
+                else {
+                    warningMessage.text("Shoot cooldown!");
+                    warningMessage.show();
+                    setTimeout(function hideMessage() { warningMessage.hide(); }, 1000)
+                }
+                break;
+            case 3:
+                //right mouse button pressed
+                var myDefense = $('#my-defense');
```
Comment (cristimihok) [inline app/static/js/game.js]: Refactor `case` statement. Extract login in functions.

File: app/static/js/game.js [hunk 26]
```diff
@@ -0,0 +1,247 @@
...
+        var timer2 = setInterval(function()
+                {
+                //mingea zboara cat timp nu iese din ecran
+                var enemyDefense = $("#enemy-defense");
+                var stop = "fly";
+                if(pro.y<gameBox.height()-pro.r && pro.x < gameBox.width())
+                {
+                myMissile.css( { display: "inline" });
+                //daca scutul e invizibil - zboare mai  departe linistit
+                if ( enemyDefense.css('display') == 'none'){
```
Comment (cristimihok) [inline app/static/js/game.js]: The state of a DOM object should be kept in a controller not in the view of the webapp.

File: app/static/js/game.js [hunk 27]
```diff
@@ -0,0 +1,247 @@
...
+                        pro.x = startX + v0x * frameCount;
+                    }
+                    //daca e vizibil si a ajuns la scut
+                    else {
+                        stop="shield";
+                        //mingea se opreste la scut daca e vizibil
+                        clearInterval(timer2);
+                    }
+                }
+                }
```
Comment (cristimihok) [inline app/static/js/game.js]: What is this closing?

File: app/static/js/game.js [hunk 28]
```diff
@@ -0,0 +1,247 @@
...
+                //mingea trece de scut daca e invizibil
+            else {
+            if(stop=="fly") {
```
Comment (cristimihok) [inline app/static/js/game.js]: Indentation

File: app/static/js/game.js [hunk 29]
```diff
@@ -0,0 +1,247 @@
...
+    }
+    
+    function displayGravity() {
+        //$("#gravity-bars:nth-child(g)").addClass("on");
+    }
+    function startTimer() {
+        var secs = 0;
+        var timer = $('#timer');
+        setInterval(function() {
+        secs+=1;
```
Comment (cristimihok) [inline app/static/js/game.js]: Indentation

File: app/static/css/game.css [hunk 30]
```diff
@@ -1,16 +1,8 @@
 .game-section {
     position: relative;
-    top: 50%;
+    top: 48%;
```
Comment (luciandragomir) [inline app/static/css/game.css]: Why is it 48 and not 50?

File: app/static/css/game.css (hunk 1 above)
Comment (luciandragomir) [inline app/static/css/game.css]: nth-child would be one approach. Oh and English please.

File: app/static/css/game.css [hunk 31]
```diff
@@ -20,3 +12,287 @@
...
+.noua { height: 81%; }
+.zece  { height: 90%; }
+.unspe { height: 99%;}
+
+.on { background: red; }
+.off { background: transparent; }
+/* end of general layout and design settings */
+
+/* current player's stats */
+#my-username {
```
Comment (luciandragomir) [inline app/static/css/game.css]: I disagree. Create a .player-username class and put there the common values, then the add the differences between the two. Same goes for the items bellow.

File: app/static/css/game.css [hunk 32]
```diff
@@ -20,3 +12,287 @@
...
+@keyframes changewidth {
+    from {
+        width: 0;
+    }
+    to {
+        width: 300px;
+    }
+}
+
+div#current-power {
```
Comment (luciandragomir) [inline app/static/css/game.css]: Why would you specify an element selector together with an id selector. An id is unique within the page.

File: app/static/js/game.js [hunk 33]
```diff
@@ -0,0 +1,347 @@
+function Game() {
```
Comment (luciandragomir) [inline app/static/js/game.js]: This goes right on the window object? Put it inside a namespace please.

File: app/static/js/game.js [hunk 34]
```diff
@@ -0,0 +1,347 @@
...
+    this.gameBox = $("#game-box");
+    this.warningMessage = $("#message");
+    this.gravity = 4.34;
+    this.timer = $("#timer");
+
+    this.displayGravity = function() {
+        //$("#gravity-bars:nth-child(g)").addClass("on");
+    }
+    this.startTimer = function(timer) {
```
Comment (luciandragomir) [inline app/static/js/game.js]: Blank space

File: app/static/js/game.js [hunk 35]
```diff
@@ -0,0 +1,347 @@
...
+        var secs = 0;
+        setInterval(function() {
```
Comment (luciandragomir) [inline app/static/js/game.js]: When are you going to cancel the interval?

File: app/static/js/game.js (hunk 33 above)
Comment (kyokidG) [inline app/static/js/game.js]: Should I make a namespace named "Game" and put all the stuff in that namespace?

File: app/static/js/game.js [hunk 36]
```diff
@@ -0,0 +1,347 @@
...
+        secs+=1;
+        if (secs<59) {
+            if (secs<10) { timer.text("0" + secs); }
+            else { timer.text(secs); }
+        }
+        else {
+            if (secs%60<10) { timer.text(Math.round(secs/60) + ":0" + secs%60); }
```
Comment (luciandragomir) [inline app/static/js/game.js]: // 110 seconds = 01:50
Math.round(110/60) = 2 // timer.text(02:50)

File: app/static/js/game.js [hunk 37]
```diff
@@ -0,0 +1,347 @@
...
+        }
+    }, 1000);}
+    this.getDatafromServer = function() {
+        //here we get the gravity and other general stuff from server
+        //gets called once in the begining of the game
+    }
+    this.initializeGame = function() {
+        $(document).bind("contextmenu", function(event) {
+        event.preventDefault();
+    });
```
Comment (luciandragomir) [inline app/static/js/game.js]: Weird indentation. The initializeGame doesn't end here, whereas the indentation suggests it does.

File: app/static/css/game.css [hunk 38]
```diff
@@ -20,3 +12,287 @@
     position: relative;
     top: 20px;
 }
+
+/* general layout and design settings */
+#game-box {
```
Comment (luciandragomir) [inline app/static/css/game.css]: There's a lot of duplicate code in this file.

File: app/static/js/game.js [hunk 39]
```diff
@@ -0,0 +1,347 @@
...
+    this.shield = $("#my-shield");
+    this.currentShield = $("#my-current-shield");
+    this.missile = $("#my-missile");
+    this.cannon = $("#my-cannon");
+    this.body = $("#my-body");
+    this.defense = $("#my-defense");
+    this.power = $("#power");
+    this.currentPower = $("#current-power");
+
+    this.isDead = function () {
```
Comment (luciandragomir) [inline app/static/js/game.js]: Does it matter if it's the current player or the enemy? The algorithm for checking if a player is dead is the same no matter which player it is. 
This is duplicate in the other Player class.
Why are there two classes for player? Shouldn't you have only one class with two (or more) instances?

File: app/static/js/game.js [hunk 40]
```diff
@@ -0,0 +1,347 @@
...
+        return this.currentHealth.width() <= 0
+    };
+    this.shot = function (startPosition, angle, power) {
```
Comment (luciandragomir) [inline app/static/js/game.js]: This method is way too big.

File: app/static/js/game.js [hunk 41]
```diff
@@ -0,0 +1,347 @@
...
+        var hit = jQuery.Deferred();
+        var pro = {
+            x: startPosition[0],
+            y: startPosition[1],
+            r: 10,
+            v: power,
+            theta: angle
+        };
+        var frameCount = 0;
+        var v0x = pro.v * Math.cos(pro.theta * Math.PI / 180);
```
Comment (luciandragomir) [inline app/static/js/game.js]: v0x is really descriptive....not!

File: app/static/js/game.js [hunk 42]
```diff
@@ -0,0 +1,347 @@
...
+    this.defense = $("#my-defense");
+    this.power = $("#power");
+    this.currentPower = $("#current-power");
+
+    this.isDead = function () {
+        return this.currentHealth.width() <= 0
+    };
+    this.shot = function (startPosition, angle, power) {
+        var hit = jQuery.Deferred();
+        var pro = {
```
Comment (luciandragomir) [inline app/static/js/game.js]: What's pro?

File: app/static/js/game.js [hunk 43]
```diff
@@ -0,0 +1,347 @@
...
+            theta: angle
+        };
+        var frameCount = 0;
+        var v0x = pro.v * Math.cos(pro.theta * Math.PI / 180);
+        var v0y = pro.v * Math.sin(pro.theta * Math.PI / 180);
+        var startX = pro.x;
+        var startY = pro.y;
+
+        var timer2 = setInterval(function () {
+            if (pro.y < game.gameBox.height() - pro.r && pro.x < game.gameBox.width()) {
```
Comment (luciandragomir) [inline app/static/js/game.js]: Too many ifs and elses. A function should have max 10 lines of code, if it's bigger, it's too big. You can split it into many small functions that do only one thing. This one does too much.

File: app/static/js/game.js [hunk 44]
```diff
@@ -0,0 +1,347 @@
...
+    this.missile = $("#enemy-missile");
+    this.cannon = $("#enemy-cannon");
+    this.body = $("#enemy-body");
+    this.defense = $("#enemy-defense");
+
+    this.isDead = function () {
+        return this.currentHealth.width() <= 0
+    };
+    this.hasDefense = function () {
+        return this.defense.css('display') != 'none'
```
Comment (luciandragomir) [inline app/static/js/game.js]: What if you later decide that a shield is always displayed, but it becomes semi-transparent?
You shouldn't rely on the presentation layer to give you application logic.
You should have a property on the object ``hasDefense`` and use that instead.

File: app/static/js/game.js [hunk 45]
```diff
@@ -0,0 +1,347 @@
...
+    }
+    this.shot = function (startPosition, angle, power) {
+        var hit = jQuery.Deferred();
```
Comment (luciandragomir) [inline app/static/js/game.js]: duplicate

File: app/static/js/game.js [hunk 46]
```diff
@@ -0,0 +1,347 @@
...
+    var startPosition = getStartPosition(positionAngle);
+    var angle = getAngle(e);
+    angle = polishAngle(angle);
+    if (isCooldownReady(currentPlayer.currentCooldown, currentPlayer.cooldown)) {
+        currentPlayer.missile.show();
+        $.when(currentPlayer.shot(startPosition, angle, power)).then(
+            function () {
+                missileHit(enemyPlayer, 101);
+                if (enemyPlayer.isDead()) {
+                    showWaringMessage("Ai castigat!");
```
Comment (luciandragomir) [inline app/static/js/game.js]: Business rules say that when a player is dead, the game ends.
When the game ends, display a message, stop timer, ...

File: app/static/js/game.js [hunk 47]
```diff
@@ -0,0 +1,347 @@
...
+
+function showWaringMessage(textMessage) {
+    game.warningMessage.text(textMessage);
+    game.warningMessage.show();
+    setTimeout(function hideMessage() {
+        game.warningMessage.hide();
+    }, 1000)
+}
+
+function getAngle(e) {
```
Comment (luciandragomir) [inline app/static/js/game.js]: Get angle of what?

File: app/static/js/game.js [hunk 48]
```diff
@@ -0,0 +1,347 @@
...
+    }
+    return angle;
+}
+
+function rotateCannon(cannon, angle) {
+    cannon.css({'transform': 'rotate(' + angle + 'deg)'});
+}
+
+function getStartPosition(angle) {
+    var xPos = currentPlayer.cannon.offset().left - 1 - game.gameBox.offset().left + 1.3 * angle;
```
Comment (luciandragomir) [inline app/static/js/game.js]: what's 1.3?

File: app/static/js/game.js [hunk 49]
```diff
@@ -0,0 +1,347 @@
...
+    var yPos = Math.floor(currentPlayer.cannon.offset().top - 1 - game.gameBox.offset().top);
+    return [xPos, yPos];
+}
+
+function getPower() {
+    var p = currentPlayer.currentPower.width() / 3;
+    currentPlayer.power.hide();
```
Comment (luciandragomir) [inline app/static/js/game.js]: Why is this side-effect here? getters should be pure functions.

File: app/static/js/game.js [hunk 50]
```diff
@@ -0,0 +1,347 @@
...
+    return p;
+}
+
+function isCooldownReady(child, wrapper) {
+    return (child.width() >= Math.floor(wrapper.width()))
+}
+
+function missileHit(enemyPlayer, damage) {
```
Comment (luciandragomir) [inline app/static/js/game.js]: Shouldn't this be on the player class? ``player.hit(damage)``

File: app/static/js/game.js [hunk 51]
```diff
@@ -0,0 +1,347 @@
...
+function shotCooldownReset(player) {
+    player.currentCooldown.css({width: 0 + "px"});
+    var shotCooldownTimer = setInterval(function () {
+        if (player.currentCooldown.width() <= Math.floor(player.cooldown.width())) {
+            player.currentCooldown.css({width: player.currentCooldown.width() + 1 + "px"});
+        }
+        else {
+            clearInterval(shotCooldownTimer);
+        }
+    }, 1000 / 40); //takes 5 seconds to refresh the shot
```
Comment (luciandragomir) [inline app/static/js/game.js]: 1000/40 = 25ms. How did you get the 5 seconds?

File: app/static/js/game.js [hunk 52]
```diff
@@ -0,0 +1,347 @@
...
+        if (player.currentShield.width() <= Math.ceil(player.shield.width())) {
+            player.currentShield.css({width: player.currentShield.width() + 1 + "px"});
+        }
+        else {
+            clearInterval(shieldCooldownTimer);
+        }
+    }, 1000 / 10); //takes 20 seconds to refresh the shot
+}
+
+
```
Comment (luciandragomir) [inline app/static/js/game.js]: What's with the extra space? There should be one and only one empty line at the end of file.
